File: DisasterServer/Entities/DTBall.cs
using DisasterServer.Maps;
using DisasterServer.Session;
using DisasterServer.State;

namespace DisasterServer.Entities;

public class DTBall : Entity
{
	private double _state;

	private bool _side = true;

	public override TcpPacket? Destroy(Server server, Game game, Map map)
	{
		return null;
	}

	public override TcpPacket? Spawn(Server server, Game game, Map map)
	{
		return null;
	}

	public override UdpPacket? Tick(Server server, Game game, Map map)
	{
		if (_side)
		{
			_state += 0.014999999664723873;
			if (_state >= 1.0)
			{
				_side = false;
			}
		}
		else
		{
			_state -= 0.014999999664723873;
			if (_state <= -1.0)
			{
				_side = true;
			}
		}
		return new UdpPacket(PacketType.SERVER_DTBALL_STATE, (float)_state);
	}
}

File: disasterserver_c/entities/DTBall.c
#include "Server.h"
#include <entities/DTBall.h>

bool dtball_tick(Server* server, Entity* entity)
{
	DTBall* ball = (DTBall*)entity;

	if (ball->side)
	{
		ball->state += 0.015 * server->delta;

		if (ball->state >= 1)
			ball->side = 0;
	}
	else
	{
		ball->state -= 0.015 * server->delta;

		if (ball->state <= -1)
			ball->side = 1;
	}

	Packet pack;
	PacketCreate(&pack, SERVER_DTBALL_STATE);
	PacketWrite(&pack, packet_writefloat, (float)ball->state);
	server_broadcast(server, &pack, false);

	return true;
}
